src/webgpu/shader/execution/robust_access.spec.ts

```diff
@@ -295,74 +382,80 @@ g.test('bufferMemory')
-          ${type.declaration};
-        };`;
-      byteSize = baseType.byteSize * type.std430Length;
-    }
-
-    // Build the test function that will do the tests.
-    testSource += `
-    uint runTest() {
-  `;
-
+    // Produce the accesses to the variable.
     for (const indexToTest of indicesToTest) {
-      // TODO check with constants too.
+      // TODO check with constants too if WGSL allows it.
       const index = `(${indexToTest}) * one`;
 
-      if (t.params.access === 'read') {
-        testSource += `
-          if(data[${index}] != ${type.zero}) {
-            return __LINE__;
-          }`;
-      } else if (t.params.access === 'write') {
-        testSource += `data[${index}] = ${type.zero};`;
-      } else {
-        testSource += `atomicAdd(data[${index}], 1);`;
+      switch (access) {
+        case 'read':
+          testFunctionSource += `
+            if(s.data[${index}] != ${type.zero}) {
+              return __LINE__;
+            }`;
+          break;
+
+        case 'write':
+          testFunctionSource += `s.data[${index}] = ${type.zero};`;
+          break;
+
+        case 'atomic':
+          testFunctionSource += `atomicAdd(s.data[${index}], 1);`;
+          break;
       }
     }
 
-    testSource += `
-      return 0;
-    }`;
+    // Check that the canaries haven't been modified
+    if (usesCanary) {
+      testFunctionSource += `
+        for (uint i = 0; i < 10; i++) {
+          if (s.startCanary[i] != 0xFFFFFFFFu) {
+            return __LINE__;
+          }
+          if (s.endCanary[i] != 0xFFFFFFFFu) {
+            return __LINE__;
+          }
+        }`;
+    }
 
-    // Create a buffer that contains zeroes in the allowed access area, and 42s everywhere else.
-    const [testBuffer, testInit] = t.device.createBufferMapped({
-      size: 512,
-      usage:
-        GPUBufferUsage.COPY_SRC |
-        GPUBufferUsage.UNIFORM |
-        GPUBufferUsage.STORAGE |
-        GPUBufferUsage.COPY_DST,
-    });
-    baseType.fillBuffer(testInit, 256, byteSize);
-    const testInitCopy = copyArrayBuffer(testInit);
-    testBuffer.unmap();
-
-    // Run the shader, accessing the buffer.
-    runShaderTest(t, GPUShaderStage.COMPUTE, testSource, [
-      { binding: 0, resource: { buffer: testBuffer, offset: 256, size: byteSize } },
-    ]);
-
-    // Check that content of the buffer outside of the allowed area didn't change.
-    t.expectSubContents(testBuffer, 0, new Uint8Array(testInitCopy.slice(0, 256)));
-    const dataEnd = 256 + byteSize;
-    t.expectSubContents(testBuffer, dataEnd, new Uint8Array(testInitCopy.slice(dataEnd, 512)));
+    // Run the test
+
+    // First aggregate the test source
+    const testSource = `
+      ${globalSource}
+
+      uint runTest() {
+        ${testFunctionSource}
+        return 0;
+      }`;
+
+    // Run it.
+    if (usesBuffer) {
+      // Create a buffer that contains zeroes in the allowed access area, and 42s everywhere else.
```

Comment:
Not new, but aren't we supposed to allow reads from anywhere in the buffer to allow Vulkan's robust access behavior?

Reply:
I'm almost 100% sure that we decided that the array accesses would be clamped.

Reply:
ugh, I can't keep anything straight anymore. We've gone through too many proposals on every single little decision in the API. I think you're right.